test_functions.py

```diff
@@ -188,27 +189,50 @@ def test_assumptions(self):
 
         assert assumption == expected
 
+    # testing analysis by grad year
+    def test_grad_analysis(self):
+
+        expected = {'2022': {
+                        'kr20': -2.333, 
+                        'idr': {1: -0.036, 2: 0.036, 3: 0.0, 4: 0.0, 5: 0.036, 6: 0.0, 7: 0.0}, 
+                        'difficulty': {1: 0.5, 2: 0.5,3: 0.0, 4: 0.0, 5: 0.5, 6: 0.0, 7: 1.0}, 
+                        'scores': [57.1, 71.4], 
+                        'average': 64.2, 
+                        'weighted_scores': [20.0, 40.0],
+                        'weighted_avg': 30.0,
+                        'exclude': [1],
+                        'diff_avg': 0.357,
+                        'idr_avg': 0.005,
+                        'num_correct': {1: 1, 2: 1, 3: 2, 4: 2, 5: 1, 6: 2, 7: 0},
+                        'assumptions': {1234: [7], 1235: [7]},
+                        },
+                    '2024': 'Invalid data - Not enough students'
+                    }
+        analysis = analyze_gradyears(self.data)['grad_year_analysis']
+
+        assert analysis == expected
+
     
     # testing with excludes
     def test_with_excludes(self):
         data = {
-            "student_list": [
```

Comment:
I replaced all single quotes with double quotes except for in the api_client.py and application.py because I'm not sure if its safe to change all of them in those files.